// typedoc-plugin-extras/dist/helpers.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.replaceTopMostTitleLink = exports.replaceDescription = exports.replaceTopMostTitle = exports.isUrl = exports.setupNewlineInFooter = exports.appendToFooter = exports.appendFavicon = exports.makeRelativeToRoot = void 0;
/**
 * Creates a relative path from a host file to a target file which is located in the root.
 * @example `modules/_index_.html` --> `../favicon.ico`
 * @param hostPath The path of the file which will contain the resulting relative path.
 * @param targetFilename The name of the target file.
 */
function makeRelativeToRoot(hostPath, targetFilename) {
    // Find separators.
    const match = hostPath.match(/[/\\]/g);
    if (!match)
        return targetFilename;
    // Create path with one '../' per separator and targetFilename at the end.
    const separatorCount = match.length;
    const parts = [...Array(separatorCount).fill('..'), targetFilename];
    return parts.join('/');
}
exports.makeRelativeToRoot = makeRelativeToRoot;
/**
 * Appends a favicon link at the end of `<head>` in the given html.
 * @param html HTML string to append to.
 * @param url URL of the favicon.
 */
function appendFavicon(html, url) {
    return html.replace('</head>', '\t' + // Some tabulation to fit the rest of the document.
        `<link rel="icon" href="${url}" />` +
        '\n' + // Push the end of <head> to the next line.
        '</head>');
}
exports.appendFavicon = appendFavicon;
/**
 * Appends a string value after "Generated using TypeDoc".
 * @param html HTML string to append to.
 * @param value A string value.
 */
function appendToFooter(html, value) {
    return html.replace(/(<p.*>Generated using.*TypeDoc.*)(<\/p>)/, '$1' + // Start of <p>
        value +
        '$2' // End of <p>
    );
}
exports.appendToFooter = appendToFooter;
/**
 * Sets up the newline in footer.
 * @param html HTML string to append to.
 */
function setupNewlineInFooter(html) {
    return html.replace(/(<p)(>Generated using.*TypeDoc)/, '$1' + // Opening of <p> tag
        ' style="line-height: 28px;"' + // Add spacing between the first line of footer and the date.
        '$2' // End of "Generated using TypeDoc"
    );
}
exports.setupNewlineInFooter = setupNewlineInFooter;
/**
 * Determines whether a string is a URL.
 * @param url The URL to check.
 */
function isUrl(url) {
    return /^https?:\/\//i.test(url);
}
exports.isUrl = isUrl;
/**
 * Replaces the top-most title text.
 * @param html HTML string to replace into.
 * @param title The new title to set.
 */
function replaceTopMostTitle(html, title) {
    return html.replace(/(<a href=")([^"]*)(" class="title">)([^<]*)(<\/a>)/, '$1' + // Start of <a>
        '$2' + // The href link
        '$3' + // The class
        title +
        '$5' // End of <a>
    );
}
exports.replaceTopMostTitle = replaceTopMostTitle;
/**
 * Replaces the meta description
 * @param html HTML string to replace into.
 * @param description The new description to set.
 */
function replaceDescription(html, description) {
    return html.replace(/(<meta name="description" content=")([^<]*)("\/>)/, '$1' + // Start of meta tag
        description.replace('"', '') + // The description (also preventing escaping the attribute)
        '$3' // end of meta tag
    );
}
exports.replaceDescription = replaceDescription;
/**
 * Replaces the top-most title link.
 * @param html HTML string to replace into.
 * @param link The new link to set.
 */
function replaceTopMostTitleLink(html, link) {
    return html.replace(/(<a href=")([^"]*)(" class="title">)([^<]*)(<\/a>)/, '$1' + // Start of <a>
        link +
        '$3' + // The class
        '$4' + // The title
        '$5' // End of <a>
    );
}
exports.replaceTopMostTitleLink = replaceTopMostTitleLink;
